### PR TITLE
Extract mapped places from <Beskrivning> 

### PR DESCRIPTION
**Pull request by mattiasostmar**
Looks for mapped places in the field <Beskrivning> in addition to <Ort, foto>. 
Result: 121 additional content categories added that would not have been added otherwise.

Also fixes a found indentation error in master that caused content_cats to be added only in rare cases.

Task: https://phabricator.wikimedia.org/T161878

### CODE REVIEW

File: create_infotexts.py [hunk 1]
```diff
@@ -336,17 +335,19 @@ def select_best_mapping_for_depicted_person(self, flipped_name):
...
-    def process_depicted_place(self, place_string, places_mapping):
+    def process_depicted_place(self, place_string, places_mapping, desc_string):
         """
         Create wikiformat depicted place string from raw input data.
         
         Populates data["depicted_place"] and interacts with data["commonscat"] and data["meta_cats"].
         
         :param place_string: string value <Ort, foto> in metadata item.
         :param places_mapping: Dictionary containing Commons:Medelhavsmuseet/batchUploads/Cypern_places
+        :param desc_string: string representing an image items field <Beskrivning>.
```
Comment (lokal-profil) [inline create_infotexts.py]: I would use similar terminology as in the description for place_string

File: create_infotexts.py [hunk 2]
```diff
@@ -374,15 +375,44 @@ def process_depicted_place(self, place_string, places_mapping):
                 place_as_wikitext = place_string
 
             # Don't forget to add the commons categories, even though only wikidata is used in depicted people field
-                if places_mapping[place_string].get('commonscat'):
-                    self.content_cats.append(places_mapping[place_string]["commonscat"])
+            if places_mapping[place_string].get('commonscat') and \
+            places_mapping[place_string].get('commonscat') != "-" and \
```
Comment (lokal-profil) [inline create_infotexts.py]: you don't need this one (you replaced "-" by `None`)

File: create_infotexts.py [hunk 3]
```diff
@@ -374,15 +375,44 @@ def process_depicted_place(self, place_string, places_mapping):
...
+            places_mapping[place_string].get('commonscat') not in self.content_cats:
```
Comment (lokal-profil) [inline create_infotexts.py]: I wouldn't worry about duplicates here but instead change [line 226](https://github.com/mattiasostmar/SMVK-Cypern_2017-01/blob/2d1e99df5cbda4f2b8ac4a29ae9115c3f647fc2b/create_infotexts.py#L226)
from:
`img_info["cats"] = img.content_cats`
to
`img_info["cats"] = list(set(img.content_cats))`

and similar for ´meta_cats`.

File: create_infotexts.py [hunk 4]
```diff
@@ -374,15 +375,44 @@ def process_depicted_place(self, place_string, places_mapping):
...
+                        place_as_wikitext = place
+
+                # Don't forget to add the commons categories, even though only wikidata is used in depicted people field
+                if places_mapping[place].get('commonscat') and \
+                    places_mapping[place].get('commonscat') != "-" and \
+                    places_mapping[place].get('commonscat') not in self.content_cats:
+                    self.content_cats.append(places_mapping[place]["commonscat"])
+
         self.data["depicted_place"] = place_as_wikitext
+        return place_as_wikitext
```
Comment (lokal-profil) [inline create_infotexts.py]: No need to `return` (you are storing it in `self.data["depicted_place"]` instead)

File: create_infotexts.py [hunk 5]
```diff
@@ -374,15 +375,44 @@ def process_depicted_place(self, place_string, places_mapping):
...
+            places_mapping[place_string].get('commonscat') not in self.content_cats:
+
+                self.content_cats.append(places_mapping[place_string]["commonscat"])
 
         else:
             self.meta_cats.append("Media_contributed_by_SMVK_without_mapped_place_value")
             place_as_wikitext = place_string
 
 
+        for place in places_mapping:
```
Comment (lokal-profil) [inline create_infotexts.py]: So a couple of questions here.
* Do we always want to run this second logic or only when we have not already found a mapped place?
** My gut feeling is the latter since the place field should be more reliable than the description.
** If the former. How do we prevent this from adding the same formatting a second time?

I'll stick some code related comments below but this one might affect whether the others are needed.

File: create_infotexts.py [hunk 6]
```diff
@@ -374,15 +375,44 @@ def process_depicted_place(self, place_string, places_mapping):
...
+            if place.lower() in desc_string.lower():
+
+                if places_mapping[place]["wikidata"] and places_mapping[place]["wikidata"] != "-":
```
Comment (lokal-profil) [inline create_infotexts.py]: just check `if places_mapping[place].get("wikidata"):` per discussion in last PR

File: create_infotexts.py [hunk 7]
```diff
@@ -374,15 +375,44 @@ def process_depicted_place(self, place_string, places_mapping):
...
+                    if place_as_wikitext: # There is already a depicted place found in field <Ort, foto>
```
Comment (lokal-profil) [inline create_infotexts.py]: Rather than having to have two separate methods depending on if a second is found it might be simpler to just use a list and do `'/'.join(...)` after the loop.

But an important question is whether it ever makes sense to have two places? (My gut feeling says this probably means one or both are wrong)

File: create_infotexts.py [hunk 8]
```diff
@@ -374,15 +375,44 @@ def process_depicted_place(self, place_string, places_mapping):
...
+                            wikidata=places_mapping[place]["wikidata"]
+                            )
+
+                else:
+                    if place_as_wikitext:
+                        place_as_wikitext += "/{}".format(place)
+                    else:
+                        place_as_wikitext = place
+
+                # Don't forget to add the commons categories, even though only wikidata is used in depicted people field
```
Comment (lokal-profil) [inline create_infotexts.py]: why does this reference depicted people?

File: create_infotexts.py [hunk 9]
```diff
@@ -374,15 +375,44 @@ def process_depicted_place(self, place_string, places_mapping):
...
+                if places_mapping[place].get('commonscat') and \
```
Comment (lokal-profil) [inline create_infotexts.py]: See comment above on only needing one of these conditions

File: create_infotexts.py [hunk 10]
```diff
@@ -17,7 +17,6 @@
 people_mapping_file = open("./people_mappings.json")
 people_mapping = json.loads(people_mapping_file.read())
 
-
```
Comment (lokal-profil) [inline create_infotexts.py]: I think this line is supposed to be kept to make linting happy.

**Review by lokal-profil — CHANGES_REQUESTED** [review]

File: create_infotexts.py (hunk 10 above)
Comment (mattiasostmar) [reply create_infotexts.py]: Jupp, done.

File: create_infotexts.py (hunk 1 above)
Comment (mattiasostmar) [reply create_infotexts.py]: Done

File: create_infotexts.py (hunk 2 above)
Comment (mattiasostmar) [reply create_infotexts.py]: A little hard to see, but I understand it that my check for "-" is unnessesary. I removed it. Done.

File: create_infotexts.py (hunk 3 above)
Comment (mattiasostmar) [reply create_infotexts.py]: Yes, nice one! ✨  
Done

File: create_infotexts.py (hunk 6 above)
Comment (mattiasostmar) [reply create_infotexts.py]: OK. Done

File: create_infotexts.py (hunk 8 above)
Comment (mattiasostmar) [reply create_infotexts.py]: Slip of thought. Altered.

File: create_infotexts.py (hunk 9 above)
Comment (mattiasostmar) [reply create_infotexts.py]: Done

File: create_infotexts.py (hunk 5 above)
Comment (mattiasostmar) [reply create_infotexts.py]: Actually - no. I moved the logic into the else-clause from <Ort, foto> above (not found)

File: create_infotexts.py (hunk 7 above)
Comment (mattiasostmar) [reply create_infotexts.py]: Since we actually don't need to run this if there is already - a more reliable - value for <Ort, foto> present I've removed the logic adding a "/" from both cases. Done.

File: create_infotexts.py (hunk 2 above)
Comment (lokal-profil) [reply create_infotexts.py]: So to clarify (no need for further change in code):
You replaced all "-" by Nan/None when you loaded the list from wiki. Therefore you will not have any "-" values here anymore.

File: create_infotexts.py [hunk 11]
```diff
@@ -375,41 +369,28 @@ def process_depicted_place(self, place_string, places_mapping, desc_string):
...
-            if place.lower() in desc_string.lower():
-
-                if places_mapping[place]["wikidata"] and places_mapping[place]["wikidata"] != "-":
-                    if place_as_wikitext: # There is already a depicted place found in field <Ort, foto>
-                        place_as_wikitext += "/{{{{city|1={wikidata}}}}}".format(
-                            wikidata=places_mapping[place]["wikidata"]
-                            )
-                    else:
+                    if places_mapping[place]["wikidata"]:
                         place_as_wikitext = "{{{{city|1={wikidata}}}}}".format(
```
Comment (lokal-profil) [inline create_infotexts.py]: Every time we get here (remember that we are looping) we will be overwriting the last value. Same is true for `place_as_wikitext = place`. I guess that instead we want to maintain a list of matches which we join in the end?

See also my [previous question](https://github.com/mattiasostmar/SMVK-Cypern_2017-01/pull/30#discussion_r110140513) is two matches actually a reason to stop not add any of them (and just add not mapped meta-cat)?

File: create_infotexts.py [hunk 12]
```diff
@@ -375,41 +369,28 @@ def process_depicted_place(self, place_string, places_mapping, desc_string):
...
-                if places_mapping[place].get('commonscat') and \
-                    places_mapping[place].get('commonscat') != "-" and \
-                    places_mapping[place].get('commonscat') not in self.content_cats:
-                    self.content_cats.append(places_mapping[place]["commonscat"])
+                    # Don't forget to add the commons categories if present.
+                    if places_mapping[place].get('commonscat'):
+                        self.content_cats.append(places_mapping[place]["commonscat"])
+
+                else:
+                    self.meta_cats.append("Media_contributed_by_SMVK_without_mapped_place_value")
```
Comment (lokal-profil) [inline create_infotexts.py]: This also gets triggered every time we go through the loop. Instead use the mentioned list of matches and only add meta_cats if the list is empty (or per above maybe if it has more than one entry)

File: create_infotexts.py [hunk 13]
```diff
@@ -375,41 +369,28 @@ def process_depicted_place(self, place_string, places_mapping, desc_string):
...
+                    # TODO: Handle images without depicted place value
 
         self.data["depicted_place"] = place_as_wikitext
         return place_as_wikitext
```
Comment (lokal-profil) [inline create_infotexts.py]: You never acted on [this comment](https://github.com/mattiasostmar/SMVK-Cypern_2017-01/pull/30#discussion_r110139278)

**Review by lokal-profil — CHANGES_REQUESTED** [review]

File: create_infotexts.py (hunk 11 above)
Comment (mattiasostmar) [reply create_infotexts.py]: I went along the road that there are only one place per image, but it is possible that both the region and e.g. the village are mentioned in the description. 

When looking at the code again I realized that if there is no value in <Ort ,foto> we just return "" instead of go looking for places mentioned in the `<Beskrivning>`which was what I intended. I will move the whole places-in-<Beskrivning> logic into that first if-clause.

I did a complete overhaul of the code since I also found other quirks such as a return value that was not intended since it updates self.data etc.

Done.

File: create_infotexts.py (hunk 12 above)
Comment (mattiasostmar) [reply create_infotexts.py]: Done.

File: create_infotexts.py (hunk 4 above)
Comment (mattiasostmar) [reply create_infotexts.py]: Done.

File: create_infotexts.py (hunk 13 above)
Comment (mattiasostmar) [reply create_infotexts.py]: Done here and done there 👯‍♂️

**Comment by lokal-profil** [comment]
I don't think such a drastic change to the logic is needed. 

The place value is always more valuable than whatever can be extracted from the description so definitely start with that and only look at the description if you don't have a place value.

Good that you noticed the `if not place_string` test in the beginning. I guess that can be dropped and instead we change `if place_string in places_mapping` to `if place_string and place_string in places_mapping`

**Comment by mattiasostmar** [comment]
Yes. Then we move the looking for places in description to the else-clause. Cleaner. Then I should also be able to drop the last `place_as_wikitext = place_string` (since there is `no place_string`) and just add the meta_cat.

Also altered so we only look in the description if there is no place_str.

File: create_infotexts.py [hunk 14]
```diff
@@ -387,7 +368,23 @@ def process_depicted_place(self, place_string, places_mapping, desc_string):
                 self.content_cats.append(places_mapping[place_string]["commonscat"])
 
         else:
-            place_as_wikitext += place_string
+            place_matches = []
+            for place in places_mapping:
+                if place.lower() in desc_string.lower():
+                    if places_mapping[place]["wikidata"]:
```
Comment (lokal-profil) [inline create_infotexts.py]: I repeat my question. Is it ever ok with multiple matches or does that indicate some kind of error?
In the loop I would only do `place_matches.append(place)` then only `if len(place_matches) == 1` I would add category and construct the string otherwise add the meta_cat.

File: create_infotexts.py [hunk 15]
```diff
@@ -387,7 +368,23 @@ def process_depicted_place(self, place_string, places_mapping, desc_string):
...
+                            wikidata=places_mapping[place]["wikidata"]))
+                    else:
+                        place_matches.append(place)
+
+                    # Don't forget to add the commons categories if present.
+                    if places_mapping[place].get('commonscat'):
+                        self.content_cats.append(places_mapping[place]["commonscat"])
+
+            if len(place_matches) == 0:
+                self.meta_cats.append("Media_contributed_by_SMVK_without_mapped_place_value")
```
Comment (lokal-profil) [inline create_infotexts.py]: Possibly we would here like to try `place_as_wikitext = place_string or ""` as a last attempt.

**Review by lokal-profil — CHANGES_REQUESTED** [review]

File: create_infotexts.py (hunk 14 above)
Comment (mattiasostmar) [reply create_infotexts.py]: I looked into it at and indeed it was due to errors.

Now it leaves only one erroneous double-hit which I'll handle manually per https://phabricator.wikimedia.org/T163669

File: create_infotexts.py (hunk 15 above)
Comment (mattiasostmar) [reply create_infotexts.py]: With the current implementation place_string will always be None if len(place_matches) == 0, so no use adding that.

Not done.

File: create_infotexts.py [hunk 16]
```diff
@@ -30,14 +30,15 @@ def load_places_mapping():
     places_df = places_df.set_index("Nyckelord")
     places_df.columns = ["freq", "commonscat", "wikidata"]
 
-    places_df.replace("-", np.nan)
+    places_df[places_df == "-"] = None
```
Comment (lokal-profil) [inline create_infotexts.py:34]: This change is a bit unclear to me.

File: create_infotexts.py [hunk 17]
```diff
@@ -236,6 +239,7 @@ class CypernImage:
 
     def __init__(self):
         """Instantiate a single instance of a processed image."""
+        self.idno = None
```
Comment (lokal-profil) [inline create_infotexts.py]: add a comment explaining (similar to the other internal variables)

File: create_infotexts.py [hunk 18]
```diff
@@ -342,47 +346,53 @@ def process_depicted_place(self, place_string, places_mapping, desc_string):
...
+                    self.meta_cats.append("Media_contributed_by_SMVK_without_mapped_place_value")
+                    self.meta_cats.append("Media_contributed_by_SMVK_possibly_depicting_Machairas_Monastary")
+                    place_as_wikitext = "Macheras"
+
+                elif places_mapping[place_string]["wikidata"]:
+                    place_as_wikitext += "{{{{city|1={wikidata}}}}}".format(
+                        wikidata=places_mapping[place_string]["wikidata"]
+                        )
+                # Don't forget to add the commons categories, even though only wikidata is used in depicted people field
+                elif places_mapping[place_string].get('commonscat'):
```
Comment (lokal-profil) [inline create_infotexts.py]: This cannot be an `elif` (should be a plain `if`).

File: create_infotexts.py [hunk 19]
```diff
@@ -342,47 +346,53 @@ def process_depicted_place(self, place_string, places_mapping, desc_string):
...
-                self.meta_cats.append("Media_contributed_by_SMVK_without_mapped_place_value")
-                self.meta_cats.append("Media_contributed_by_SMVK_possibly_depicting_Machairas_Monastary")
-                place_as_wikitext = "Macheras"
-
-            elif places_mapping[place_string]["wikidata"]:
-                place_as_wikitext += "{{{{city|1={wikidata}}}}}".format(
-                    wikidata=places_mapping[place_string]["wikidata"]
-                )
+        if place_string:
+            if place_string in places_mapping:
```
Comment (lokal-profil) [inline create_infotexts.py]: The reshuffling means that any `place_string` which is mapped but does not have a _wikidata_ entry does not get any `place_as_wikitext` (before it would get `place_string` in that case). Is this change on purpose? Does this situation ever arise?

File: create_infotexts.py [hunk 20]
```diff
@@ -342,47 +346,53 @@ def process_depicted_place(self, place_string, places_mapping, desc_string):
...
             else:
                 place_as_wikitext += place_string
 
-            # Don't forget to add the commons categories, even though only wikidata is used in depicted people field
-            if places_mapping[place_string].get('commonscat'):
-                self.content_cats.append(places_mapping[place_string]["commonscat"])
-
         else:
             place_matches = []
+            place_cnt = 0
```
Comment (lokal-profil) [inline create_infotexts.py]: remove `place_cnt` since you never use it

File: create_infotexts.py (hunk 14 above)
Comment (lokal-profil) [reply create_infotexts.py]: You are still handling this as though multiple entries are expected.

Simply use something like the below (from L375, so all i indented once)
```python
place_matches = []
for place in places_mapping:
    if place.lower() in desc_string.lower():
        place_matches.append(place)

if len(place_matches) == 1:
    if places_mapping[place]["wikidata"]:
        place_as_wikitext = {{{{city|1={wikidata}}}}}".format(wikidata=places_mapping[place]["wikidata"]))
    else:
        place_as_wikitext = place

    if places_mapping[place].get('commonscat'):
        self.content_cats.append(places_mapping[place]["commonscat"])
else:
   self.meta_cats.append("Media_contributed_by_SMVK_without_mapped_place_value"

    if len(place_matches) > 1:
        print("Found {} places in description! -> {} img: {}".format(len(place_matches), place_matches, self.idno))
```

**Comment by lokal-profil** [comment]
The last review comment ended up at https://github.com/mattiasostmar/SMVK-Cypern_2017-01/pull/30#discussion_r114004750

File: create_infotexts.py (hunk 16 above)
Comment (mattiasostmar) [reply create_infotexts.py:34]: Reasonably so! I noticed that np.nan leaves the string "nan" in the fields which makes our boolean tests further down the line not to work! Not happy with Pandas due to this - it took me a long time to find out.

File: create_infotexts.py (hunk 17 above)
Comment (mattiasostmar) [reply create_infotexts.py]: Done

File: create_infotexts.py (hunk 19 above)
Comment (mattiasostmar) [reply create_infotexts.py]: No, It's supposed to get the place value as a string. It actually happens in two cases. Done.

File: create_infotexts.py (hunk 18 above)
Comment (mattiasostmar) [reply create_infotexts.py]: True! Done.

File: create_infotexts.py (hunk 20 above)
Comment (mattiasostmar) [reply create_infotexts.py]: Done - also had to remove it's non-functional a few lines down.

File: create_infotexts.py (hunk 14 above)
Comment (mattiasostmar) [reply create_infotexts.py]: Yes, it was good during debugging (I found a few wrong mappings), but this is cleaner and a good end result. Done.

File: create_infotexts.py (hunk 16 above)
Comment (lokal-profil) [reply create_infotexts.py:34]: I'm not sure this will actually work though.

What you want to do is go through every entry and if the value is `"-"` then replace it by `None`. Here it looks like you are adding a new entry to the dict (with the key being `False` or `True`) and setting it's value to None. I.e. you are not replacing the `"-"` as you intend.

**Review by lokal-profil — CHANGES_REQUESTED** [review]
Only https://github.com/mattiasostmar/SMVK-Cypern_2017-01/pull/30#discussion_r114495704 left

File: create_infotexts.py (hunk 16 above)
Comment (mattiasostmar) [reply create_infotexts.py:34]: It's the way pandas works: I use a [boolean filter](http://pandas.pydata.org/pandas-docs/stable/indexing.html#boolean-indexing) that says that each field in the entire DataFrame/data-dict where the value is '== "-"' should be set to None. I see by the results that it works.

Comment (lokal-profil) [reply create_infotexts.py:34]: OK. Me being confused about it being a panda object.

Then just fix the typo on L29 ("**is** a dataframe") and insert a blank line after L23

Comment (mattiasostmar) [reply create_infotexts.py:34]: Yes, good - the Pandas DataFrame is actually just intermediary to simplify the reading of Excel-files. Clarified in docstring.
Done.

**Review by lokal-profil — APPROVED** [review]
